---
title: "01_FilterAndTrim.Rmd"
author: "Gabriel Tedone"
date: "2024-11-07"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, root.dir = "/home/ecosystems/Bioinformatics/Classes/BioInfo4EEG/Your_Project_Folder")
```

# Filtering and Trimming of .fastq reads

[**Note:**]{.underline} This step comes after **demultiplexing**, a step that splits a big **.fastq** file containing all reads of a single sequencing run (with reads comming from different samples all mixed in together) into **separate .fastq files**, each one containing sample-specific reads and/or either forward/reverse reads. Demultiplexing can be done in multiple ways (QIIME2, bash tools, etc). However, this processing step is typically done by the sequencing facility, so it will not be covered on this small course. For more info on this step (demultiplexing), visit [this demultiplexing tutorial (tool: Sabre)](https://astrobiomike.github.io/amplicon/demultiplexing){.uri} or ...

## Overview

There is an ubiquous **filtering** and **trimming** process applied to DNA/RNA sequencing reads (.fastq files) that aims at obtaining clean .fastq reads, by elimination, trimming and filtering of **raw sequencing data**. *You can think of this process as a way of eliminating "noise" and obtaining a clearer signal than the raw output by the sequencing machine.*

Each read is analysed for it's **quality**. This is informed by the overall read length as well as by the quality of each nucleotide (in Casava .fastq format this is informed by the **Phred score**) and **primers** are removed (**trimmed**).

In this course, we will focus the **quality analysis** on reads produced by **Illumina** **pair-ended sequencing machines (Mi-Seq, etc)**.

### Tools

**FastQC** (*bash CLI*): **Q**uality **C**ontrol of reads in a **.fastq** file ([link](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/){.uri}).

![](images/fastqc.png)

![](images/goodQualityfastqc.png)

**MultiQC** (*bash CLI*): combines multiple fastqc quality analysis into one interactive, easy to read .html file ([link](https://seqera.io/multiqc/)).

![](images/multiqc.png)

**Cutadapt** (*python*): will search for and trim primer sequences in the reads, as well as trim based on nucleotide quality, length, etc. [Note:]{.underline} Although the cutadapt tool runs on *python*, in this tutorial it will be implemented within an *R script.*

### Steps

1.  Initial Quality Check -\> *FastQC + MultiQC*

2.  Primer Trimming -\> *Cutadapt* (*within* *R)*

3.  Read Quality Trimming and Filtering -\> *dada2::filterAndTrim()* \~ similar to cutadapt

## Code

[Note:]{.underline} the following commands are optimized for use on the **EEG bioinformatic's linux machine**, since all the necessary programs are already installed and working. If working on another machine: set-up a conda environment with the necessary python version and the programs (fastqc, multiqc and cutadapt), like shown here [link].

### 1. Initial QUALITY Check (on bash)

```{bash}
#[bash CLI on EEG's bionformatics machine]

# Listing the available conda environments
conda env list
```

```{bash}
# Activating the 'metatax' conda environment
conda activate metatax

```

```{bash}
fastqc --help # have a quick look at the tool's parameters

fastqc -o [PATH] [.fastq.gz files] 

        
```

[Note:]{.underline} Although this step is marked sequentially as a first step, it is a good practice to keep on checking on read's quality and length as we go on the filtering and trimming process. Visualizing and interpreting the read quality plots will ensure the process is working as intended, and it will help deciding filtering thresholds.

#### Start of the R script

```{r warning=TRUE}
#!/usr/bin/Rscript

# - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - #

cat("This script will take in .fastq files and: 
1) remove indeterminations (Ns) 
2) trim primers used for barcode selection & amplification
3) filter and trim low quality nucleotide reads
\n")
cat("Working directory:", getwd(), "\n\n")

# - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - #

# Libraries ----
library(Biostrings) # fast manipulation of large biological sequences or sets of sequences
library(ShortRead) # sampling, iteration, and input of FASTQ files. filtering and trimming reads, 
                   # quality assessment report.
library(dplyr) # data management
library(dada2)

# - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - #
```

```{r}
# - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - #

# Set general parameters ----
setwd("/home/ecosystems/Bioinformatics/Classes/BioInfo4EEG/Your_Project_Folder")

amplicon <- "16S"
cat(paste("Amplicon:", amplicon), "\n")
cat("Inputfiles' path: \n")
inpath <- paste("", amplicon, sep = "/"); print(inpath); cat("\n")
cat("Outputfiles' path: \n")
outpath <- paste("data/microbial_community/01_Trimming_and_Filtering", amplicon, sep = "/"); print(outpath); cat("\n")

# PRIMERS [CHANGE ME!]
cat("Forward and Reverse PRIMERS: \n")
FWD <- "CCTACGGGNGGCWGCAG"; print(FWD)
REV <- "GACTACHVGGGTATCTAATCC"; print(REV)
cat("\n")

# - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - # - #

# Input data ----

# .fastq files overview
print("We are going to analyse the following files:")
print(list.files(inpath, pattern = ".fastq.gz"))
cat("\n")
print(paste("NÂº of files:", length(list.files(inpath, pattern = ".fastq.gz")))) ; cat("\n")

# Save forward and reverse reads file paths
fnFs <- sort(list.files(inpath, 
                        pattern = "_R1_001.fastq.gz", # [change me!] Forward filename pattern
                        full.names = TRUE)) 
fnRs <- sort(list.files(inpath, 
                        pattern = "_R2_001.fastq.gz", # [change me!] Reverse filename pattern
                        full.names = TRUE)) 

```

### 2. PRIMER Trimming with Cutadapt

In this step, we will specify the **primer's** **sequences** (both *forward* and *reverse*). These sequences are typically known via prior knowledge during DNA extraction/library preparation and are usually specified within the sequencing report.

The following script will parse all sequences and count the number of primers found. This "primer check" will also be done after running cutadapt, to see if, in fact, primers were removed. Because of how PCR works, we will have to look for **reverse**, **complement** and **reverse complement** sequences of primers.

```{r}

```

### 3. Read Quality Filtering and Trimming

```{r}

```
