---
title: "00_FilterAndTrim_Sequences"
author: "Gabriel Tedone"
date: "2024-10-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

# Filtering and Trimming Reads

**Note:** This step comes after **demultiplexing**, a step that splits a big .fastq file containing all reads of a single sequencing run (with reads comming from different samples all mixed in together) into separate .fastq files, each one containing sample-specific reads and/or either forward/reverse reads. Demultiplexing can be done in multiple ways (QIIME2, bash tools, etc). However, this processing step is typically done by the sequencing facility, so it will not be covered on this small course. For more info on this step (demultiplexing), visit <https://astrobiomike.github.io/amplicon/demultiplexing> or

There is a ubiquous filtering and trimming process applied to DNA/RNA sequencing reads (.fastq files) that considers modifications to DNA fragments added during library preparation

```{r}
getwd()
list.files()
```
